web/js/codeworld_shared.js

```diff
@@ -122,13 +122,111 @@ var hintBlacklist = [
     "Pi",
 ];
 
-var codeWorldBuiltinsDocs = {};
+
+// codeWorldDocs is variable containing annotations and documentation
+// of builtin and user-defined variables.
+// Expected format:
+// codeWorldDocs = {
+//   codeWorldLogo: {
+//     annotation: "Picture",
+//     doc: "The CodeWorld logo."
+//   }
+// }
+
+var codeWorldDocs = {};
+
+function renderHover(keyword) {
+    var elem = document.createElement('div')
+
+    if (!codeWorldDocs[keyword]){
+        return null;
+    };
+    elem.title = keyword;
+    var keywordData = codeWorldDocs[keyword];
+
+    // Variable name + annotaion
+    var topLine = document.createElement("div");
+    topLine.className = "hover-decl";
+    var wordElem = document.createElement("span");
+    wordElem.className = "hint-word";
+    wordElem.appendChild(document.createTextNode(keyword));
+    topLine.appendChild(wordElem);
+    if (keywordData.annotation) {
+        var annotation = document.createElement('div');
+        topLine.appendChild(document.createTextNode(" :: " + keywordData.annotation))
+    };
+    elem.appendChild(topLine)
+
+    // Docs
+    if (keywordData.doc) {
+        var docElem = document.createElement('div');
+        docElem.innerHTML = keywordData.doc;
+        docElem.className = "hover-doc";
+        elem.appendChild(docElem)
+    };
+    return elem;
+};
+
+function applyCodemirrorChange(line, change) {
```

Comment:
This is worrisome.  Does this mean that CodeMirror will silently ignore any kind of edit except for insertions and deletions?  Skimming the documentation, it looks like at least "+move" is an option, and others besides.  I'm pretty convinced this is the wrong way to go, as it's likely to be a minefield that will break with every editor feature you haven't quite tested yet.

What's the purpose behind ignoring `setValue` and `complete`?

Reply:
In this case this mean that CodeMirror will ignore all except insertions and deletions in [beforeChange handler](https://github.com/google/codeworld/pull/799/files#diff-0b4870211bd48634500a9dccbf7798dfR173). 
But anyway, `applyCodemirrorChange` exactly not finished yet. If it will be used in other places it could break something. `complete` ignored because have no use (I could be wrong) in updating doc dictionary. If some term could be completed - it already in documentation.

About `setValue`. `setValue` happens on loading project, so yes, it is error. All keywords in fresh loaded project will not be parsed.

Also, `applyCodeMirrorChange` works only per line now, I guess there is actions which could touch  more than one.

 I will think about all this.